File: src/main/java/no/skatteetaten/aurora/databasehotel/metrics/ResourceUseCollector.java
package no.skatteetaten.aurora.databasehotel.metrics;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.MeterRegistry;
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema;
import no.skatteetaten.aurora.databasehotel.service.DatabaseHotelService;
import no.skatteetaten.aurora.databasehotel.utils.MapUtils;

@Component
public class ResourceUseCollector {

    private static final Logger LOG = LoggerFactory.getLogger(ResourceUseCollector.class);
    private static final String METRIC_NAME = "aurora.dbh.schema.size.mb";
    private static final String ENVIRONMENT_LABEL = "environment";
    private static final String APP_LABEL = "application";
    private static final String AFFILIATION_LABEL = "affiliation";
    private static final String ENVIRONMENT_METRIC_LABEL = "namespace";
    private final MeterRegistry registry;
    private final DatabaseHotelService databaseHotelService;
    private final Map<String, SchemaSizeValue> schemaSizeValues = new HashMap<>();

    public ResourceUseCollector(DatabaseHotelService databaseHotelService, MeterRegistry registry) {
        this.registry = registry;
        this.databaseHotelService = databaseHotelService;
    }

    @Scheduled(fixedDelayString = "${metrics.resourceUseCollectInterval}", initialDelay = 15000)
    public void collectResourceUseMetrics() {

        long start = System.currentTimeMillis();
        LOG.info("Collecting resource use metrics");

        Set<DatabaseSchema> allSchemas = databaseHotelService.findAllDatabaseSchemas(null);
        LOG.debug("Found {} schemas total", allSchemas.size());

        List<DatabaseSchema> databaseSchemas = allSchemas.stream()
            .filter(s -> MapUtils.containsEveryKey(s.getLabels(), ENVIRONMENT_LABEL, APP_LABEL, AFFILIATION_LABEL))
            .collect(Collectors.toList());

        // Let's update the SchemaSizeValues that backs the schema size gauges
        databaseSchemas.forEach(schema -> {
            SchemaSizeValue existingValue = schemaSizeValues.get(schema.getId());
            if (existingValue != null) {
                existingValue.setSizeMb(schema.getSizeMb());
            } else {
                SchemaSizeValue value = new SchemaSizeValue(schema.getId(), schema.getSizeMb());
                schemaSizeValues.put(schema.getId(), value);
                registerGaugeForSchema(schema, value);
            }
        });

        // Remove the SchemaSizeValues that does not currently have a DatabaseSchema object. This indicates that
        // the DatabaseSchema has been deleted since the last time we updated the gauges. Deleting the value that
        // backs a Gauge *should* also in fact delete the Gauge since the reference to the SchemaSizeValue is a
        // WeakReference. This does not appear to happen (bug?(!)), and to make matters worse, deleting the Gauge
        // from the MeterRegistry does not delete it from the io.prometheus.client.CollectorRegistry instance that is
        // used to render the data for Prometheus. I'm hoping and assuming that these issues will be fixed in a future
        // version of micrometer.io, so I'm leaving the implementation as is - even if that means leaving deleted
        // DatabaseSchemas with a NaN size.
        schemaSizeValues.keySet().removeIf(
            schemaId -> {
                boolean isAnExistingSchema =
                    databaseSchemas.stream().anyMatch(databaseSchema -> schemaId.equals(databaseSchema.getId()));
                boolean isSchemaRemovedSinceLastIteration = !isAnExistingSchema;
                return isSchemaRemovedSinceLastIteration;
            });

        LOG.info("Resource use metrics collected for {} schemas in {} millis", databaseSchemas.size(),
            System.currentTimeMillis() - start);
    }

    private void registerGaugeForSchema(DatabaseSchema schema, SchemaSizeValue value) {

        Map<String, String> labels = schema.getLabels();
        Gauge.builder(METRIC_NAME, value, SchemaSizeValue::getSizeMb)
            .description("the size of the individual registered schemas")
            .tags(createLabelsArray(labels))
            .register(registry);
    }

    private String[] createLabelsArray(Map<String, String> labels) {
        return new String[] { AFFILIATION_LABEL, labels.get(AFFILIATION_LABEL),
            APP_LABEL, labels.get(APP_LABEL),
            ENVIRONMENT_METRIC_LABEL, labels.get(ENVIRONMENT_LABEL) };
    }

    public static class SchemaSizeValue {
        private String id;

        private double sizeMb;

        public SchemaSizeValue(String id, double sizeMb) {
            this.id = id;
            this.sizeMb = sizeMb;
        }

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public double getSizeMb() {
            return sizeMb;
        }

        public void setSizeMb(double sizeMb) {
            this.sizeMb = sizeMb;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            SchemaSizeValue that = (SchemaSizeValue) o;

            return id.equals(that.id);
        }

        @Override
        public int hashCode() {
            return id.hashCode();
        }
    }
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/metrics/ResourceUseCollector.kt
package no.skatteetaten.aurora.databasehotel.metrics

import io.micrometer.core.instrument.Gauge
import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Tag
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.AFFILIATION
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.APPLICATION
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.DATABASE_ENGINE
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.DATABASE_HOST
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.ENVIRONMENT
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.INSTANCE_NAME
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.NAMESPACE
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.SCHEMA_ID
import no.skatteetaten.aurora.databasehotel.metrics.MetricTag.SCHEMA_TYPE
import no.skatteetaten.aurora.databasehotel.service.DatabaseHotelService
import no.skatteetaten.aurora.databasehotel.service.DatabaseInstance
import no.skatteetaten.aurora.databasehotel.service.TablespaceInfo
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.util.HashMap

@Component
class ResourceUseCollector(
    private val databaseHotelService: DatabaseHotelService,
    private val registry: MeterRegistry
) {
    private val schemaGauges = mutableMapOf<String, Pair<GaugeValue, Gauge>>()
    private val schemaCountGauges = mutableMapOf<String, Pair<GaugeValue, Gauge>>()
    private val availibleTablespacesGauges = mutableMapOf<String, Pair<GaugeValue, Gauge>>()

    @Scheduled(fixedDelayString = "\${metrics.resourceUseCollectInterval}", initialDelay = 5000)
    fun collectResourceUseMetrics() {

        LOG.info("Collecting resource use metrics")
        val start = System.currentTimeMillis()

        val databaseSchemas = loadDatabaseSchemas()
        val tablespaceInfo = databaseHotelService.getTablespaceInfo()

        handleSchemaSizeMetrics(databaseSchemas)
        handleSchemaCountMetrics(databaseSchemas)
        handleAvailibleTablespacesMetrics(tablespaceInfo)

        LOG.info(
            "Resource use metrics collected for {} schemas in {} millis", databaseSchemas.size,
            System.currentTimeMillis() - start
        )
    }

    private fun loadDatabaseSchemas() =
        databaseHotelService.findAllDatabaseSchemas(null, HashMap(), true)
            .toList()
            .also { LOG.debug("Found {} schemas total", it.size) }

    private fun handleSchemaSizeMetrics(databaseSchemas: List<DatabaseSchema>) {
        val currentMetricIds = databaseSchemas.map { schema ->
            val id = schema.id
            val existing = schemaGauges[id]
            if (existing != null) {
                val metricData = existing.first
                metricData.value = schema.sizeMb * 1024 * 1024
            } else {
                val data = GaugeValue(id, schema.sizeMb)
                val gauge = registerSchemaSizeGauge(schema, data)
                schemaGauges[id] = data to gauge
            }
            id
        }

        schemaGauges.removeDeprecatedMetrics(currentMetricIds)
    }

    private fun handleSchemaCountMetrics(databaseSchemas: List<DatabaseSchema>) {
        val grouped: Map<CountGroup, List<DatabaseSchema>> = groupDatabaseSchemasForCounting(databaseSchemas)
        val currentMetricIds: List<String> = grouped.map { (groupKeys, schemas) ->
            val id = groupKeys.id
            val existing = schemaCountGauges[id]
            if (existing != null) {
                val metricData = existing.first
                metricData.value = schemas.size.toDouble()
            } else {
                val data = GaugeValue(id, schemas.size.toDouble())
                val gauge = registerSchemaCountGauge(groupKeys, data)
                schemaCountGauges[id] = data to gauge
            }
            id
        }

        schemaCountGauges.removeDeprecatedMetrics(currentMetricIds)
    }

    private fun handleAvailibleTablespacesMetrics(tablespaceInfo: List<Pair<DatabaseInstance, TablespaceInfo>>) {
        val currentMetricIds: List<String> = tablespaceInfo.map { (instance, tablespaceInfo) ->
            val id = instance.metaInfo.host
            val existing = availibleTablespacesGauges[id]
            if (existing != null) {
                val metricData = existing.first
                metricData.value = tablespaceInfo.available.toDouble()
            } else {
                val data = GaugeValue(id, tablespaceInfo.available.toDouble())
                val gauge = registerAvailibleTablespacesGauge(instance, data)
                availibleTablespacesGauges[id] = data to gauge
            }
            id
        }
        availibleTablespacesGauges.removeDeprecatedMetrics(currentMetricIds)
    }

    private fun groupDatabaseSchemasForCounting(databaseSchemas: List<DatabaseSchema>) =
        databaseSchemas.groupBy {
            mapOf(
                "state" to if (it.active) "ACTIVE" else "COOLDOWN",
                "databaseHost" to it.databaseInstanceMetaInfo.host,
                "databaseEngine" to it.databaseInstanceMetaInfo.engine.name,
                "schemaType" to it.type.name,
                "affiliation" to it.labels[AFFILIATION_LABEL]
            )
        }

    private fun registerSchemaSizeGauge(schema: DatabaseSchema, value: GaugeValue): Gauge {
        return Gauge.builder(SCHEMA_SIZE_METRIC_NAME, value, { it.value })
            .baseUnit("bytes")
            .description("the size of the schema")
            .tags(schema.metricTags)
            .register(registry)
    }

    private fun registerSchemaCountGauge(groupKeys: CountGroup, value: GaugeValue): Gauge {
        val tags = groupKeys.map { (k, v) -> tagOfValueOrUnknown(k, v) }
        return Gauge.builder(SCHEMA_COUNT_METRIC_NAME, value, { it.value })
            .baseUnit("count")
            .description("The amount of schemas")
            .tags(tags)
            .register(registry)
    }

    private fun registerAvailibleTablespacesGauge(databaseInstance: DatabaseInstance, value: GaugeValue): Gauge {
        return Gauge.builder(AVAILIBLE_TABLESPACES_METRIC_NAME, value, { it.value })
            .baseUnit("availible")
            .description("The amount of availible tablespaces left on databaseHost")
            .tags(databaseInstance.metricTags)
            .register(registry)
    }

    private fun MutableMap<String, Pair<GaugeValue, Gauge>>.removeDeprecatedMetrics(currentMetricIds: List<String>) {
        val metricsToRemove = keys.minus(currentMetricIds)
        metricsToRemove
            .mapNotNull { this[it] }
            .forEach { (data, gauge) ->
                remove(data.id)
                registry.remove(gauge.id)
            }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(ResourceUseCollector::class.java)
        private const val SCHEMA_SIZE_METRIC_NAME = "aurora.dbh.schema.size.bytes"
        private const val SCHEMA_COUNT_METRIC_NAME = "aurora.dbh.schema.count"
        private const val AVAILIBLE_TABLESPACES_METRIC_NAME = "aurora.dbh.availible.tablespaces"
    }
}

private enum class MetricTag(val tagName: String) {
    SCHEMA_ID("id"),
    SCHEMA_TYPE("schemaType"),
    DATABASE_HOST("databaseHost"),
    DATABASE_ENGINE("databaseEngine"),
    AFFILIATION("affiliation"),
    ENVIRONMENT("environment"),
    APPLICATION("application"),
    NAMESPACE("namespace"),
    INSTANCE_NAME("instanceName")
}

private val NAMESPACE_LABEL = ENVIRONMENT.tagName
private val APP_LABEL = APPLICATION.tagName
private val AFFILIATION_LABEL = AFFILIATION.tagName

private val DatabaseSchema.metricTags
    get(): List<Tag> {
        val labels = this.labels
        val t = { tag: MetricTag, value: String? -> tagOfValueOrUnknown(tag.tagName, value) }
        val namespace = labels[NAMESPACE_LABEL]?.removeSuffix("-")
        val affiliation = labels[AFFILIATION_LABEL]
        val environment = affiliation?.let { namespace?.getEnvironment(it) }
        return listOf(
            t(SCHEMA_ID, this.id),
            t(AFFILIATION, affiliation),
            t(APPLICATION, labels[APP_LABEL]),
            t(ENVIRONMENT, environment),
            t(NAMESPACE, namespace),
            t(DATABASE_HOST, this.databaseInstanceMetaInfo.host),
            t(DATABASE_ENGINE, this.databaseInstanceMetaInfo.engine.name),
            t(SCHEMA_TYPE, this.type.name)
        )
    }

private val DatabaseInstance.metricTags
    get(): List<Tag> {
        val t = {tag: MetricTag, value: String? -> tagOfValueOrUnknown(tag.tagName,value)}
        return listOf(
            t(DATABASE_HOST, this.metaInfo.host),
            t(INSTANCE_NAME,  this.instanceName),
            t(DATABASE_ENGINE, this.metaInfo.engine.name)
        )
    }

data class GaugeValue(var id: String, var value: Double)

private fun tagOfValueOrUnknown(name: String, value: String?) = Tag.of(name, value ?: "UNKNOWN")

private typealias CountGroup = Map<String, String?>

private val CountGroup.id get() = this.toSortedMap().map { (k, v) -> "$k=$v" }.joinToString(",")

private fun String.getEnvironment(affiliation: String): String {
    val environment = this.replace("^$affiliation".toRegex(), "").removePrefix("-")
    return if (environment.isBlank()) affiliation else environment
}
